### PR TITLE
IRGen: Make resilient enum's tag indices resilient

### PR DESCRIPTION
**Pull request by aschwaighofer**
This allows reordering enum cases resiliently.

rdar://24057946

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
Please add a test to validation-test/Evolution that shows this working at run time! :-)

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/IRGen/GenEnum.cpp [hunk 1]
```diff
@@ -6107,6 +6105,15 @@ void IRGenModule::emitEnumDecl(EnumDecl *theEnum) {
   }
 
   emitFieldMetadataRecord(theEnum);
+
+  if (isResilient(theEnum, ResilienceExpansion::Maximal))
```
Comment (slavapestov) [inline lib/IRGen/GenEnum.cpp]: Don't you want to return if its *not* resilient?

**Comment by slavapestov** [comment]
Can you file a bug to have me or Huon update TBDGen as well? Or you can do it, it's trivial. You just have to emit the link entities when visiting an EnumDecl, and add a test case.

File: lib/IRGen/GenEnum.cpp (hunk 1 above)
Comment (aschwaighofer) [reply lib/IRGen/GenEnum.cpp]: The code below emits the global variables for resilient enum tags. I only want to proceed to this code if we are within the same resilience domain not outside.

**Comment by aschwaighofer** [comment]
I can take a look

**Comment by aschwaighofer** [comment]
Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test source compatibility

**Comment by aschwaighofer** [comment]
Successful source compatibility run is here: https://ci.swift.org/view/Pull%20Request/job/swift-PR-source-compat-suite/914/console

File: lib/IRGen/GenEnum.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/IRGen/GenEnum.cpp]: When are we ever resilient here? Isn’t this the code for emitting an enum _definition__? How can you define an enum that’s resilient to yourself?

Comment (aschwaighofer) [reply lib/IRGen/GenEnum.cpp]: Good point. Rethinking that logic should we test if the enum is resilient and only then emit the globals (like you initially suggested)? Assuming, the transition from non-resilient to resilient is not allowed.

Comment (aschwaighofer) [reply lib/IRGen/GenEnum.cpp]: Like so: https://github.com/apple/swift/pull/15401

**Pull request merged**